libparsec/crates/client/src/workspace/merge.rs

```diff
@@ -193,21 +284,52 @@ pub(super) fn merge_local_folder_manifest(
         return MergeLocalFolderManifestOutcome::NoChange;
     }
 
-    let local_from_remote = LocalFolderManifest::from_remote_with_local_context(
-        remote,
-        prevent_sync_pattern,
-        local,
-        timestamp,
-    );
+    // TODO: The next step is rarely actually needed, we could optimize this
+    //       by passing a `force_apply_pattern` parameter to this function.
 
-    // 2) Shortcut in case only the remote has changed
+    // 2) Ensure the prevent sync pattern is applied (idempotent)
+    //
+    // The prevent sync pattern can change at any time, which in turn triggers
+    // a full refresh of all local manifest to update their confinement points.
+    // Here we make sure the local manifest doesn't have its confinement points
+    // outdated, which would cause inconsistency in the next step.
+    let local = &local.apply_prevent_sync_pattern(prevent_sync_pattern, timestamp);
+
+    // 3) Confinement points is a special case: given any entry that match the prevent
+    // sync pattern is kept appart, there is no possibility of conflict here (i.e.
+    // confined local entries stay in `local.children`, confined remote entries stay in
+    // `local.base.children`).
+    //
+    // So if the current local manifest only changes are confined entries, then the merge
+    // operation only consist of converting the remote into a local and adding those
+    // confined entries back.
+    // This is precisely what we are doing in this step, and why the resulting manifest
+    // is called `merge_in_progress`: if the local manifest contains other changes, then
+    // we must do a proper children merge and update the manifest accordingly.
+
+    let mut merge_in_progress =
+        LocalFolderManifest::from_remote_with_restored_local_confinement_points(
+            remote,
+            prevent_sync_pattern,
+            local,
+            timestamp,
+        );
+    let remote = &merge_in_progress.base;
+
+    // 4) Shortcut in case only the remote has changed
     if !local_need_sync {
-        return MergeLocalFolderManifestOutcome::Merged(Arc::new(local_from_remote));
+        // Note a weird corner case here:
+        // If the prevent sync pattern has changed but the local manifest hasn't had
+        // time to be updated yet, then it's possible to end up with
+        // `local_need_sync == false`, but `merge_in_progress.need_sync == true` !
+
+        // TODO: determine if that's an issue !!!
```

Comment:
Is there a github issue ?

Reply:
yes https://github.com/Scille/parsec-cloud/issues/7924